[vst3sdk] New port for Steinberg Virtual Studio Technology 3 SDK

JoergAtGithub opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [n.a.] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [n.a.] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

JoergAtGithub committed:
New port vst3sdk

JoergAtGithub commented:
The optional GUI module https://github.com/steinbergmedia/vstgui is not added to this port, as it has own version numbering and release cycle.

I was not able to get the macOS builds working on CI and therefore disabled them. It seems that the projects CMake files require the GUI version of Xcode and do not work with the command line version.

dg0yt commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,135 @@
+# Therefore these are not part of this port
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
+FEATURES
+    "plugin-examples"            SMTG_ENABLE_VST3_PLUGIN_EXAMPLES
+    "hosting-examples"           SMTG_ENABLE_VST3_HOSTING_EXAMPLES
+    "audiounit-wrapper"          SMTG_ENABLE_AUV2_BUILDS
```
```suggestion
    FEATURES
        "plugin-examples"        SMTG_ENABLE_VST3_PLUGIN_EXAMPLES
        "hosting-examples"       SMTG_ENABLE_VST3_HOSTING_EXAMPLES
        "audiounit-wrapper"      SMTG_ENABLE_AUV2_BUILDS

```

dg0yt commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,135 @@
+
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}"
+    OPTIONS -DSMTG_ENABLE_VSTGUI_SUPPORT=OFF
+	        -DSMTG_CREATE_PLUGIN_LINK=OFF
+	${FEATURE_OPTIONS}
```
```suggestion
    OPTIONS
        -DSMTG_ENABLE_VSTGUI_SUPPORT=OFF
	    -DSMTG_CREATE_PLUGIN_LINK=OFF
        ${FEATURE_OPTIONS}

```

dg0yt replied:
Please don't indent with tab.

dg0yt commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,135 @@
+	${FEATURE_OPTIONS}
+)
+
+vcpkg_build_cmake(
+    SOURCE_PATH "${SOURCE_PATH}"
+)
```
```suggestion
vcpkg_cmake_build()
```
(if not ..._install)

dg0yt reviewed (commented):
Regarding osx:
- Not `supported` seems wrong, it is more a CI problem, i.e. a `fail` in the CI baseline.
- Did you try to set `ENV{XCODE_VERSION}` and `ENV{SDKROOT}`?
  (IDK why this project wants environment variables at this point.)
  https://github.com/steinbergmedia/vst3_cmake/blob/2022b265cdbcd90b37294d3fb397f9766dbe4c06/modules/SMTG_DetectPlatform.cmake#L48-L60

JoergAtGithub commented:
I already tried to set ENV{XCODE_VERSION}, but this alone was not sufficient.

dg0yt commented:
Taking another look, that cmake script doesn't look convincing at all.
They determine (cmake) `XCODE_VERSION` when `ENV{XCODE_VERSION}` isn't set.
They determine (cmake) `CMAKE_OSX_SYSROOT` when `ENV{SDKROOT}` isn't set.

The key question is if they really need the XCode generator or GUI at some point, or if they just didn't know how to to make their cmake build more flexible.

JoergAtGithub committed:
Formating and code style suggestions

JoergAtGithub committed:
Add osx as supported by the port, but set `arm64-osx` and `x64-osx` as failing in CI baseline

dg0yt commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,134 @@
+        ${FEATURE_OPTIONS}
+)
+
+vcpkg_build_cmake()
```
Once again:
```suggestion
vcpkg_cmake_build()
```

JoergAtGithub committed:
Use vcpkg_cmake_build() instead of vcpkg_build_cmake()

JonLiu1993 commented on versions/v-/vst3sdk.json:
```diff
@@ -0,0 +1,14 @@
+      "git-tree": "600ed441266d9a9efb240aeb1474a42f218fc44f",
+      "version-string": "v3.7.12_build_20",
+      "port-version": 0
+    },
+    {
+      "git-tree": "060f1781b84540bba8f47edeaaa1979e544aad15",
+      "version": "3.7.12",
+      "port-version": 0
```
```suggestion
```

JonLiu1993 commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,134 @@
+)
+
+# Suppress CRT linkage check
+set(VCPKG_POLICY_SKIP_CRT_LINKAGE_CHECK enabled)
```
```suggestion
set(VCPKG_POLICY_SKIP_CRT_LINKAGE_CHECK enabled)

```

JonLiu1993 replied:
You need to add a blank line at the end

JoergAtGithub committed:
Added missing empty line at portfile end
Removed accidental added version entry "3.7.12" from vst3sdk.json.

dg0yt commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,134 @@
+    "${CURRENT_PACKAGES_DIR}/include/vst3sdk/public.sdk/source/vst/utility/test"
+)
+
+# Suppress CRT linkage check
+set(VCPKG_POLICY_SKIP_CRT_LINKAGE_CHECK enabled)
```
Why is it *okay* ot suppress this check?

(Why doesn't it use `vcpkg_check_linkage()`?)

JoergAtGithub replied:
I could not get it to work with `vcpkg_check_linkage` so I removed these triplets from the list of supported triplets by `!staticcrt`

JonLiu1993 commented:
/azp run

JonLiu1993 commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,134 @@
+   endif()
+endif()
+
+file(INSTALL "${SOURCE_PATH}/LICENSE.txt" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)
```
```suggestion
vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/LICENSE.txt")
```

dg0yt commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,134 @@
+file(INSTALL "${SOURCE_PATH}/pluginterfaces/vst/" DESTINATION "${CURRENT_PACKAGES_DIR}/include/vst3sdk/pluginterfaces/vst/" FILES_MATCHING PATTERN "*.h")
+file(INSTALL "${SOURCE_PATH}/public.sdk/source/vst/" DESTINATION "${CURRENT_PACKAGES_DIR}/include/vst3sdk/public.sdk//source/vst/" FILES_MATCHING PATTERN "*.h")
+
+if (VCPKG_TARGET_IS_LINUX)
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/bin/Debug/moduleinfotool" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/bin/Debug/validator" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   else()
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/bin/moduleinfotool${VCPKG_TARGET_EXECUTABLE_SUFFIX}" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/bin/validator${VCPKG_TARGET_EXECUTABLE_SUFFIX}" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   endif()
```
Why choose the debug binary? (This breaks release-only triplets.)

dg0yt commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,134 @@
+endif()
+
+if ("hosting-examples" IN_LIST FEATURES)
+if (VCPKG_TARGET_IS_LINUX)
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/bin/Debug/editorhost" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   else()
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-dbg/bin/editorhost${VCPKG_TARGET_EXECUTABLE_SUFFIX}" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   endif()
```
Why choose the debug binary? (This breaks release-only triplets.)

dg0yt reviewed (commented):
Please format if/else like
~~~cmake
if(...)
    if(...)
        a()
    else()
        b()
    endif()
endif()

JoergAtGithub committed:
Use vcpkg_install_copyright
Use executables from release build, to support release-only community triplets
Fixed if/else/endif indention

JonLiu1993 commented:
@JoergAtGithub, could you please reply @dg0yt's review?

JoergAtGithub committed:
Removed VCPKG_POLICY_SKIP_CRT_LINKAGE_CHECK and removed staticcrt triplets from suported list

JonLiu1993 commented:
Tested features [`hosting-examples,hosting-examples`] successfully in the followng triplet:
- x86-windows
- x64-windows
- x64-windows-static

vicroms commented on ports/vst3sdk/portfile.cmake:
```diff
@@ -0,0 +1,133 @@
+file(INSTALL "${SOURCE_PATH}/pluginterfaces/vst/" DESTINATION "${CURRENT_PACKAGES_DIR}/include/vst3sdk/pluginterfaces/vst/" FILES_MATCHING PATTERN "*.h")
+file(INSTALL "${SOURCE_PATH}/public.sdk/source/vst/" DESTINATION "${CURRENT_PACKAGES_DIR}/include/vst3sdk/public.sdk//source/vst/" FILES_MATCHING PATTERN "*.h")
+
+if (VCPKG_TARGET_IS_LINUX)
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/bin/Release/moduleinfotool" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/bin/Release/validator" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+else()
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/bin/moduleinfotool${VCPKG_TARGET_EXECUTABLE_SUFFIX}" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+   file(INSTALL "${CURRENT_BUILDTREES_DIR}/${TARGET_TRIPLET}-rel/bin/validator${VCPKG_TARGET_EXECUTABLE_SUFFIX}" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}")
+endif()
```
Are these executables required for the proper function of the library? Executable files should be installed in `tools\${PORT}` rather than `share\${PORT}`

JoergAtGithub replied:
Done!

JoergAtGithub committed:
Install executable in tools\${PORT} instead of share\${PORT}
